### PR TITLE
Add thumbcache plugin

### PR DESCRIPTION
**Pull request by Miauwkeru**

### CODE REVIEW

File: dissect/target/plugins/os/windows/thumbcache.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            if cache_path.exists():
+                yield cache_path
+
+    def check_compatible(self) -> None:
+        for cache_path in self.get_cache_paths():
+            if len(list(cache_path.glob("*_idx.db"))) > 0:
+                return None
+        raise UnsupportedPluginError("There was no cache path for that plugin")
+
+    def plugin(self, record_type: TargetRecordDescriptor, prefix: str):
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/windows/thumbcache.py:48]: Return type?

File: dissect/target/plugins/os/windows/thumbcache.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                        flags=index_entry.flags,
+                        last_modified=index_entry.last_modified,
+                        path=str(cache.index_file),
+                    )
+            except Exception as e:
+                self.target.log.error(e)
+                pass
+
+    @export(record=ThumbcacheRecord)
+    def thumbcache(self):
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/windows/thumbcache.py:76]: Return type?

File: dissect/target/plugins/os/windows/thumbcache.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        yield from self.plugin(ThumbcacheRecord, "thumbcache")
+
+    @export(record=IconcacheRecord)
+    def iconcache(self):
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/windows/thumbcache.py:80]: Return type?

File: dissect/target/plugins/os/windows/thumbcache.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    def get_cache_paths(self) -> Iterator[TargetPath]:
+        for user_details in self.target.user_details.all_with_home():
+            cache_path = user_details.home_path / "appdata/local/microsoft/windows/explorer"
+            if cache_path.exists():
+                yield cache_path
+
+    def check_compatible(self) -> None:
+        for cache_path in self.get_cache_paths():
+            if len(list(cache_path.glob("*_idx.db"))) > 0:
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/windows/thumbcache.py:44]: You can reduce this code with 1 line if you use:

```
if len(list(cache_path.glob("*_idx.db"))) == 0:
   raise UnsupportedPluginError("There was no cache path for that plugin")
```

File: dissect/target/plugins/os/windows/thumbcache.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                for index_entry in cache.index_entries():
+                    yield IndexRecord(
+                        identifier=index_entry.identifier.hex(),
+                        in_use=index_entry.in_use(),
+                        flags=index_entry.flags,
+                        last_modified=index_entry.last_modified,
+                        path=str(cache.index_file),
+                    )
+            except Exception as e:
+                self.target.log.error(e)
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/windows/thumbcache.py:72]: Should there not be an additional explanation in this error message or is just the exception enough?

**Review by cecinestpasunepipe — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/thumbcache.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    [
+        ("string", "identifier"),
+        ("boolean", "in_use"),
+        ("varint", "flags"),
+        ("datetime", "last_modified"),
+        ("uri", "path"),
+    ],
+)
+ThumbcacheRecord = TargetRecordDescriptor("thumbcache/thumbcache", GENERIC_THUMBCACHE_FIELDS)
+IconcacheRecord = TargetRecordDescriptor("thumbcache/iconcache", GENERIC_THUMBCACHE_FIELDS)
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/thumbcache.py:30]:
```suggestion
ThumbcacheRecord = TargetRecordDescriptor("windows/thumbcache/thumbcache", GENERIC_THUMBCACHE_FIELDS)
IconcacheRecord = TargetRecordDescriptor("windows/thumbcache/iconcache", GENERIC_THUMBCACHE_FIELDS)
```

File: dissect/target/plugins/os/windows/thumbcache.py (hunk 1 above)
Comment (Schamper) [inline dissect/target/plugins/os/windows/thumbcache.py:48]: `plugin` is maybe a bit too generic of a function name. Maybe `_parse` or something?

**Review by Schamper — CHANGES_REQUESTED** [review]

File: setup.py [hunk 7]
```diff
@@ -75,7 +75,7 @@ def get_args_patched(cls, dist, header=None):
             "dissect.fat>=3.0.dev,<4.0.dev",
             "dissect.ffs>=3.0.dev,<4.0.dev",
             "dissect.sql>=3.0.dev,<4.0.dev",
-            "dissect.xfs>=3.0.dev,<4.0.dev",
+            "dissect.thumbcache>=0.0.dev,<1.0.dev" "dissect.xfs>=3.0.dev,<4.0.dev",
```
Comment (Schamper) [inline setup.py:78]:
```suggestion
            "dissect.thumbcache>=0.0.dev,<1.0.dev",
            "dissect.xfs>=3.0.dev,<4.0.dev",
```

File: dissect/target/plugins/os/windows/thumbcache.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    ("string", "identifier"),
+    ("string", "hash"),
+    ("string", "extension"),
+    ("varint", "data_size"),
+    ("bytes", "header_checksum"),
+    ("bytes", "data_checksum"),
+    ("uri", "path"),
+]
+IndexRecord = TargetRecordDescriptor(
+    "thumbcache/index",
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/thumbcache.py:20]:
```suggestion
    "windows/thumbcache/index",
```

File: dissect/target/plugins/os/windows/thumbcache.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            except Error as e:
+                # A specific thumbcache exception occurred, log the error.
+                self.target.log.error(e)
+            except Exception as e:
+                # A different exception occurred, log the exception.
+                self.target.log.critical(e, exc_info=True)
+                pass
+
+    @export(record=ThumbcacheRecord)
+    def thumbcache(self) -> Iterator[ThumbcacheRecord]:
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/thumbcache.py:82]: Maybe add a `-o`/`--output` argument to extract the images? If that argument is given, extract the images to that directory.

**Review by cecinestpasunepipe — APPROVED** [review]

**Pull request merged**